### PR TITLE
Shrink Session a bit

### PR DESCRIPTION
**Pull request by bjorn3**
Remove a couple of unnecessary fields from `Session` and remove a `Lock<T>` for a field that is never mutated anyway.

### CODE REVIEW

File: compiler/rustc_session/src/session.rs [hunk 1]
```diff
@@ -196,7 +190,7 @@ pub struct Session {
     /// Tracks the current behavior of the CTFE engine when an error occurs.
     /// Options range from returning the error without a backtrace to returning an error
     /// and immediately printing the backtrace to stderr.
-    pub ctfe_backtrace: Lock<CtfeBacktrace>,
```
Comment (Aaron1011) [inline compiler/rustc_session/src/session.rs:199]: This is modified by Miri

Comment (bjorn3) [reply compiler/rustc_session/src/session.rs]: I see. Removed this commit.

Comment (RalfJung) [reply compiler/rustc_session/src/session.rs]: Good catch, @Aaron1011 !

Might be good to add a comment explaining this in the code.

Comment (bjorn3) [reply compiler/rustc_session/src/session.rs:199]: Added

File: compiler/rustc_codegen_cranelift/src/driver/aot.rs [hunk 2]
```diff
@@ -68,14 +68,13 @@ fn reuse_workproduct_for_cgu(
     cgu: &CodegenUnit<'_>,
     work_products: &mut FxHashMap<WorkProductId, WorkProduct>,
 ) -> CompiledModule {
-    let incr_comp_session_dir = tcx.sess.incr_comp_session_dir();
```
Comment (cjgillot) [inline compiler/rustc_codegen_cranelift/src/driver/aot.rs:71]: Is something keeping the handling of the incr-comp session directory from being removed from Session?

Comment (bjorn3) [reply compiler/rustc_codegen_cranelift/src/driver/aot.rs:71]: It has to be stored somewhere. `TyCtxt` won't easily work as cg_ssa and cg_llvm require `sess.incr_comp_session_dir()` at a place where `TyCtxt` is not available. It is probably possible to refactor this away from `Session` and I was already looking into it a bit, but then I got distracted by refactorings to the crate loader.

**Comment by bjorn3** [comment]
Spurious? This didn't touch rustdoc.

**Comment by GuillaumeGomez** [comment]
There are no flaky in those tests (they only check the generated HTML). Not sure what broke in your PR though... Maybe rebase and try again?

**Comment by bjorn3** [comment]
Rebased.

@bors r=cjgillot

**Comment by pnkfelix** [comment]
Oddly this was flagged as causing a [large regression in instruction counts](https://perf.rust-lang.org/compare.html?start=64929313f53181636e4dd37e25836973205477e4&end=fcce644119cf4e8e36001368e514bb5ed67cb855&stat=instructions:u) for deeply-nested-async benchmark during perf triage.

**Comment by GuillaumeGomez** [comment]
Wild guess would be that it comes from [this line](https://github.com/rust-lang/rust/pull/88530/files#diff-a83e0054300d480e1f72a82b5b49c0f01c5f4e49475b0ed2427074d14c67f82cR890).

**Comment by bjorn3** [comment]
https://github.com/rust-lang/rust/pull/88748 will revert that change.

**Comment by Mark-Simulacrum** [comment]
> It turns out that the regression was in an unrelated area of the compiler and fixed after splitting the mir crate, before revert even landed. (#89322)

So, marking as triaged.

**Pull request merged**